overlord/snapstate/autorefresh_gating_test.go

```diff
@@ -0,0 +1,392 @@
+	defer restore()
+
+	st := s.state
+
+	st.Lock()
+	defer st.Unlock()
+	kernel := s.mockInstalledSnap(c, kernelYaml, false)
+	s.mockInstalledSnap(c, snapCyaml, true)
+	s.mockInstalledSnap(c, snapByaml, false)
+
+	updates := []*snap.Info{kernel}
+	affected, err := snapstate.AffectedByRefresh(st, updates)
+	c.Assert(err, IsNil)
+	c.Check(affected, DeepEquals, map[string]*snapstate.AffectedSnapInfo{
+		"snap-c": {
+			Restart: true,
+			AffectingSnaps: map[string]bool{
+				"kernel": true,
+			}}})
+}
+
+func (s *autorefreshGatingSuite) TestAffectedByGadget(c *C) {
+	restore := release.MockOnClassic(true)
+	defer restore()
+
+	st := s.state
+
+	st.Lock()
+	defer st.Unlock()
+	kernel := s.mockInstalledSnap(c, gadget1Yaml, false)
+	s.mockInstalledSnap(c, snapCyaml, true)
+	s.mockInstalledSnap(c, snapByaml, false)
+
+	updates := []*snap.Info{kernel}
+	affected, err := snapstate.AffectedByRefresh(st, updates)
+	c.Assert(err, IsNil)
+	c.Check(affected, DeepEquals, map[string]*snapstate.AffectedSnapInfo{
+		"snap-c": {
+			Restart: true,
+			AffectingSnaps: map[string]bool{
+				"gadget": true,
+			}}})
+}
+
+func (s *autorefreshGatingSuite) TestAffectedBySlot(c *C) {
+	restore := release.MockOnClassic(true)
+	defer restore()
+
+	st := s.state
+
+	st.Lock()
+	defer st.Unlock()
+
+	snapD := s.mockInstalledSnap(c, snapDyaml, true)
+	snapE := s.mockInstalledSnap(c, snapEyaml, true)
+	// unrelated snap
+	s.mockInstalledSnap(c, snapFyaml, true)
+
+	c.Assert(s.repo.AddSnap(snapD), IsNil)
+	c.Assert(s.repo.AddSnap(snapE), IsNil)
+	cref := &interfaces.ConnRef{PlugRef: interfaces.PlugRef{Snap: "snap-e", Name: "plug"}, SlotRef: interfaces.SlotRef{Snap: "snap-d", Name: "slot"}}
+	_, err := s.repo.Connect(cref, nil, nil, nil, nil, nil)
+	c.Assert(err, IsNil)
+
+	updates := []*snap.Info{snapD}
+	affected, err := snapstate.AffectedByRefresh(st, updates)
+	c.Assert(err, IsNil)
+	c.Check(affected, DeepEquals, map[string]*snapstate.AffectedSnapInfo{
+		"snap-e": {
+			AffectingSnaps: map[string]bool{
+				"snap-d": true,
+			}}})
+}
+
+func (s *autorefreshGatingSuite) TestNotAffectedByCoreOrSnapdSlot(c *C) {
+	restore := release.MockOnClassic(true)
+	defer restore()
+
+	st := s.state
+
+	st.Lock()
+	defer st.Unlock()
+
+	snapE := s.mockInstalledSnap(c, snapEyaml, true)
+	core := s.mockInstalledSnap(c, coreYaml, false)
+	snapd := s.mockInstalledSnap(c, snapdYaml, false)
+	s.mockInstalledSnap(c, snapByaml, true)
+
+	c.Assert(s.repo.AddSnap(snapE), IsNil)
+	c.Assert(s.repo.AddSnap(core), IsNil)
+
+	cref := &interfaces.ConnRef{PlugRef: interfaces.PlugRef{Snap: "snap-e", Name: "plug"}, SlotRef: interfaces.SlotRef{Snap: "core", Name: "slot"}}
+	_, err := s.repo.Connect(cref, nil, nil, nil, nil, nil)
+	c.Assert(err, IsNil)
+
+	updates := []*snap.Info{core, snapd}
+	affected, err := snapstate.AffectedByRefresh(st, updates)
+	c.Assert(err, IsNil)
+	c.Check(affected, HasLen, 0)
+}
+
+func (s *autorefreshGatingSuite) TestAffectedByPlugWithMountBackend(c *C) {
```

Comment:
we need to test this also for something like the desktop interface where the slot snap is the system snap which means could be snapd or core

Reply:
There is TestNotAffectedByCoreOrSnapdSlot test above, or do you mean something more specific?

Reply:
no, that test shows that Core and Snapd are ignored (NotAffected),   but the cases I mention here they would be affecting from the slot side,   what I'm referring to is a snap that has a desktop plug, desktop uses the mount backend so either core or snapd as system snap would be listed as affecting. That's what I think we want an explicit test for...